numpy/typing/_array_like.py

```diff
@@ -30,6 +45,24 @@ def __array__(self, dtype: None = ...) -> ndarray[Any, _DType]: ...
 else:
     _SupportsArray = Any
 
+# TODO: Wait for support for recursive types
+_NestedSequence = Union[
+    _T,
+    Sequence[_T],
+    Sequence[Sequence[_T]],
+    Sequence[Sequence[Sequence[_T]]],
+    Sequence[Sequence[Sequence[Sequence[_T]]]],
+]
+_RecursiveSequence = Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]]
+
+# A union representing array-like objects; consists of two typevars:
+# One representing types that can be parametrized w.r.t. `np.dtype`
+# and another one for the rest
+_ArrayLike = Union[
```

Comment:
Ideally we'd only need a single typevar here so that we can just use, _e.g._, `_ArrayLike[np.float64]`. 
Unfortunately this would require us to somehow express a `np.generic` via its' builtin counterpart (_e.g._ expressing `float` as `SuperFancyProtocol[np.float64]`), which is simply not possible as of the moment. Not without some plugin magic at least.